linera-chain/src/data_types.rs

```diff
@@ -587,19 +587,70 @@ impl<'a> SignatureAggregator<'a> {
     }
 }
 
+fn is_strictly_ordered(values: &[(ValidatorName, Signature)]) -> bool {
+    values.windows(2).all(|pair| pair[0].0 < pair[1].0)
+}
+
+impl<'de> Deserialize<'de> for Certificate {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Debug, Deserialize)]
+        #[serde(rename = "Certificate")]
+        struct CertificateHelper {
+            value: HashedValue,
+            round: RoundNumber,
+            signatures: Vec<(ValidatorName, Signature)>,
+        }
+
+        let helper: CertificateHelper = Deserialize::deserialize(deserializer)?;
+        if !is_strictly_ordered(&helper.signatures) {
+            Err(serde::de::Error::custom("Vector is not strictly sorted"))
+        } else {
+            Ok(Self {
+                value: helper.value,
+                round: helper.round,
+                signatures: helper.signatures,
+            })
+        }
+    }
+}
+
 impl Certificate {
     pub fn new(
         value: HashedValue,
         round: RoundNumber,
-        signatures: Vec<(ValidatorName, Signature)>,
+        mut signatures: Vec<(ValidatorName, Signature)>,
     ) -> Self {
+        if !is_strictly_ordered(&signatures) {
+            // Not enforcing no duplicates
+            signatures.sort_by_key(|&(validator_name, _)| validator_name)
+        }
+
         Self {
             value,
             round,
             signatures,
         }
     }
 
+    pub fn signatures(&self) -> &Vec<(ValidatorName, Signature)> {
+        &self.signatures
+    }
+
+    pub fn add_signature(
```

Comment:
Maybe also add to this method's documentation that it's the caller's responsibility to not insert duplicates.

Basically the `Ok` case in the binary search below is (or should be) unreachable.